README
Business-Edifact-Interchange

Module to parse Library Supply Edifact Messages
Following the EDItEUR Specifications

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Business::Edifact::Interchange

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-Edifact-Interchange

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Business-Edifact-Interchange

    CPAN Ratings
        http://cpanratings.perl.org/d/Business-Edifact-Interchange

    Search CPAN
        http://search.cpan.org/dist/Business-Edifact-Interchange/

    Github Repo
        git://github.com/colinsc/Edifact-Interchange.git



LICENSE AND COPYRIGHT

Copyright (C) 2011,2012 Colin Campbell <colinsc@cpan.org>

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Business-Edifact-Interchange

0.01    2011-03-04/11:00
        First version, parsing quotes messages

0.02    2012-03-18/15:00
        Placed under Business namespace

0.03    2012-04-16/16:36
        Include basic handling of Invoice Messages

0.04    2013-04-23/10:46
        Price needs to return an array of prices not
        a single price for invoices

0.05    2013-12-16/11:45
        Switch is now experimental
        Avoid errors in newer perls by declaring its use

0.06    2013-12-19/13:15
        As switch warnings now propagate abandon it
        Support for free text fields at message level
         


lib/Business/Edifact/Interchange.pm
package Business::Edifact::Interchange;

use warnings;
use strict;
use 5.010;
use Carp;
use Encode;
use Business::Edifact::Message;

=head1 NAME

Business::Edifact::Interchange - Parse Edifact Messages For Book Ordering

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

#  UNOA and UNOB "correspond to the basic ascii sets of iso 646 and iso 6937"
# Version 4 of edifact should extend this to unicode
my %encoding_map = (
    'UNOA' => 'ascii',
    'UNOB' => 'ascii',
    'UNOC' => 'iso-8859-1',
    'UNOD' => 'iso-8859-2',
    'UNOE' => 'iso-8859-5',
    'UNOF' => 'iso-8859-7',
);

=head1 SYNOPSIS

This is a support module for EDI ordering modules being developed for the
Koha and Evergreen OS Library Management Systems

    use Business::Edifact::Interchange;

    my $foo = Business::Edifact::Interchange->new();
    $foo->parse($edifact_message);
    or
    $foo->parse_file($filename);
    ...

The standards for using Edifact in Library Book Supply are available from

    www.editeur.org


=head1 SUBROUTINES/METHODS

=head2 new

Create an Business::Edifact::Interchange object

=cut

sub new {
    my $class       = shift;
    my $print_trace = shift;

    my $self = {};

    bless $self, $class;
    return $self;
}

=head2 parse

parse the edifact interchange passed in the message

=cut

sub parse {
    my $self = shift;
    my $doc  = shift;
    $self->{separator} = {
        component => q{\:},
        data      => q{\+},
        decimal   => q{.},
        release   => q{\?},
        reserved  => q{ },
        segment   => q{\'},
    };
    $self->{sep_class} = '\:\+\'\?';
    if ( $doc =~ s/^UNA// ) {    # optional
        my $element = substr $doc, 0, 6, q{};
        $self->read_service_string_advice($element);
    }
    my @segments =
      split /(?<!$self->{separator}->{release})$self->{separator}->{segment}/,
      $doc;
    $self->{interchange} = [];
    $self->{messages}    = [];
    $self->{msg_cnt}     = 0;
    if ( $segments[0] =~ m/^UNB/ ) {

        my $hdr = shift @segments;
        my @hdr_fields =
          split /(?<!$self->{separator}->{release})$self->{separator}->{data}/,
          $hdr;
        push @{ $self->{interchange} },
          $self->interchange_header( @hdr_fields[ 1 .. $#hdr_fields ] );
    }
    else {
        croak 'Interchange does not begin with an Interchange header';
    }
    my $current_msg;
    while ( my $segment = shift @segments ) {
        my ( $tag, @data ) =
          split /(?<!$self->{separator}->{release})$self->{separator}->{data}/,
          $segment;
        if ( $tag =~ /UNH/ ) {
            $current_msg = $self->message_header(@data);
            next;
        }
        if ( $tag =~ /UNT/ ) {
            $self->message_trailer( $current_msg, @data );
            $current_msg = undef;
            next;
        }
        if ( $tag =~ /^UNZ/ ) {
            $self->interchange_trailer(@data);
            next;
        }
        if ( $tag =~ /^UNG/ ) {
            $self->message_group_header(@data);
            next;
        }
        if ( $tag =~ /^UNE/ ) {
            $self->message_group_trailer(@data);
            next;
        }

        $self->user_data_segment( $current_msg, $tag, @data );
    }
    return;
}

=head2 parse_file

Reads an edifact message from a file and parses it
Will strip the lineendings added to files by some suppliers

=cut

sub parse_file {
    my $self     = shift;
    my $filename = shift;
    $filename ||= 'SampleQuote3.txt';

    open my $fh, '<', $filename or croak "Cannot open $filename : $!";
    my @lines = <$fh>;
    close $fh;

    for (@lines) {
        chomp;
        s/\r$//;
    }
    my $msg;
    if ( @lines == 1 ) {
        $msg = $lines[0];
    }
    elsif ( @lines > 1 ) {
        $msg = join q{}, @lines;
    }
    if ($msg) {
        $self->parse($msg);
    }
    return;
}

=head2 user_data_segment

internal method for handling message data segments
pass to the current Business::Edifact::Message object for fuller passing

=cut

sub user_data_segment {
    my ( $self, $msg, $tag, @data ) = @_;

    my $d = $self->split_components(@data);

    # un release data
    $msg->add_segment( $tag, $d );

    return;
}

=head2 message_header

create a new Business::Edifact::Message object

=cut

sub message_header {
    my ( $self, @data ) = @_;

    my $msg = Business::Edifact::Message->new( $self->split_components(@data) );

    return $msg;
}

=head2 message_trailer

End message add completed message to my messages array

=cut

sub message_trailer {
    my ( $self, $msg, @data ) = @_;

    $self->{msg_cnt}++;
    push @{ $self->{messages} }, $msg;
    return;
}

=head2 interchange_trailer

internal method to parse and validate the
interchange trailer

=cut

sub interchange_trailer {
    my ( $self, @data ) = @_;

    if ( $data[0] != $self->{msg_cnt} ) {
        carp "Message count error trailer says $data[0] I counted "
          . $self->{msg_cnt};
    }
    say "Ref:$data[1]";
    if ( $data[1] ne $self->{control_ref} ) {

        carp 'Error mismatched control refs Header:'
          . $self->{control_ref}
          . ' Trailer:'
          . $data[1];
    }

    return;
}

=head2 read_service_string_advice

internal method to parse the service string advice
and set the separator values for the interchange
accordingly 

=cut

sub read_service_string_advice {
    my $self = shift;
    my $ssa  = shift;

    # The six characters represent
    # component data element separator :
    # Data element sep                 +
    # Decimal notation                 .
    # Release Indicator                ?
    # Reseved (space)
    # Segment terminator               '
    if ( $ssa eq q{:+.? '} ) {

        # 'Standard Service String Advice';
        return;
    }
    else {

        #'Non standard Service String Advice';
        my @char = unpack 'C6', $ssa;
        foreach (@char) {
            $_ = quotemeta $_;
        }
        $self->{separator} = {
            component => $char[0],
            data      => $char[1],
            decimal   => $char[2],
            release   => $char[3],
            reserved  => $char[4],
            segment   => $char[5],
        };
        $self->{sep_class} = join q{}, $char[0], $char[1], $char[3], $char[5];
    }
    return;
}

=head2 split_components

internal method to split data field into components

=cut

sub split_components {
    my ( $self, @data ) = @_;
    my $d_arr = [];
    for my $data_field (@data) {
        my @components =
          split
          /(?<!$self->{separator}->{release})$self->{separator}->{component}/,
          $data_field;
        foreach (@components) {
            s/$self->{separator}->{release}([$self->{sep_class}])/$1/g;

            # convert data to utf-8
            $_ = $self->{enc}->decode($_);
        }
        push @{$d_arr}, \@components;
    }
    return $d_arr;
}

=head2 interchange_header

Internal method to parse the interchange header

=cut

sub interchange_header {
    my ( $self, @hdr ) = @_;
    $self->{control_ref} = $hdr[4];
    my $charencoding = 'iso-8859-1';
    my $syntax_id = substr $hdr[0], 0, 4;
    if ( exists $encoding_map{$syntax_id} ) {
        $charencoding = $encoding_map{$syntax_id};
    }
    $self->{enc} = find_encoding($charencoding);
    croak qq(encoding "$charencoding" not found) unless ref $self->{enc};
    my $interchange_header = $self->split_components(@hdr);

    # syntax identifier :: Syntax_id a4 'UNO'[ABC]:Syntax_version
    # interchange_sender :: Sender_id:id_code_qualifier[:Address]
    # interchange_recipient :: Recepient_id:code_qualifier[:Address]
    # DateTime of Prep : YYMMDD:HHMM
    # Interchange Control Ref
    # Password
    # Application Ref
    # [Priority Code]
    # [Ack Request]
    # [Comm Agreement ID]
    # [Test Indicator [ 1 == a test]]
    return $interchange_header;
}

=head2 message_group_header

internal method to parse the message group header
(Currently a nop )

=cut

sub message_group_header {
    my ( $self, @data ) = @_;

    #TBD parse data
    #say 'message_group_header';

    return;
}

=head2 message_group_trailer

internal method to parse the message group trailer
(Currently a nop )

=cut

sub message_group_trailer {
    my ( $self, @data ) = @_;

    #TBD parse data
    #say 'message_group_trailer';

    return;
}

=head2 messages

Returns and array_ref of Edifact::Message objects representing
the contents of the interchange

=cut

sub messages {
    my $self = shift;
    if ( exists $self->{messages} ) {
        return $self->{messages};
    }
    return;
}

=head1 WARNINGS

At present this is tested for quotes. Beware suppliers' interpretation of the
Edifact Standard can vary considerably. (And the standard is large enough to
allow considerable leeway on this). Its intended to expand this module based on
practical experience.

=head1 AUTHOR

Colin Campbell, C<< <colinsc@cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-edifact-interchange at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Business-Edifact-Interchange>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Business::Edifact::Interchange


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011,2012 Colin Campbell. <colin.campbell@ptfs-europe.com>

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;    # End of Business::Edifact::Interchange


lib/Business/Edifact/Message/LineItem.pm
package Business::Edifact::Message::LineItem;

use warnings;
use strict;
use 5.010;
use Carp;

=head1 NAME

Business::Edifact::Message::LineItem - Model an individual Item Line in a message

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

An edifact Message can caontain a number of line items
This is an instance of one

=head1 SUBROUTINES/METHODS

=head2 new

Called by Business::Edifact::Message to instantiate a new LineItem
object. The caller passes the lineitem fields to
the constructor

=cut

sub new {
    my $class = shift;
    my $self  = shift;

    bless $self, $class;
    return $self;
}

=head2 addsegment

adds a segment arrayref

=cut

sub addsegment {
    my $self        = shift;
    my $datalabel   = shift;
    my $data_arrref = shift;

    if ( !exists $self->{$datalabel} ) {
        $self->{$datalabel} = $data_arrref;
    }
    else {
        if ( ref $self->{$datalabel} eq 'ARRAY' ) {
            push @{ $self->{$datalabel} }, $data_arrref;
        }
    }
    return $self;
}

=head2 order_reference_number

=cut

sub order_reference_number {
    my $self = shift;

}

=head2 line_sequence_number

=cut

sub line_sequence_number {
    my $self = shift;
    return $self->{line_number};
}

=head2 ean

Return the lineitem's ean (a 13 digit ISBN)

=cut

sub ean {    #LIN
    my $self = shift;
}

=head2 author_surname

=cut

sub author_surname {    #010
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '010' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 author_firstname

=cut

sub author_firstname {    # 011
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '011' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 author

=cut

sub author {
    my $self = shift;
}

=head2 title

=cut

sub title {    #050
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '050' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 subtitle

=cut

sub subtitle {    #060
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '060' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 edition

=cut

sub edition {    # IMD 100
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '100' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 place_of_publication

=cut

sub place_of_publication {    # IMD 110
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '110' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 publisher

=cut

sub publisher {    # IMD 120
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '120' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 date_of_publication

=cut

sub date_of_publication {    # IMD 170
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '170' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 item_format

=cut

sub item_format {    #IMD 220
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '220' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 shelfmark

=cut

sub shelfmark {    #IMD 230
    my $self = shift;
    for my $d ( @{ $self->{item_description} } ) {
        if ( $d->{code} eq '230' ) {
            return $d->{text};
        }
    }
    return q{};
}

=head2 quantity

=cut

sub quantity {
    my $self = shift;
}

=head2 price

=cut

sub price {
    my $self = shift;
}

=head2 related_numbers

=cut

sub related_numbers {
    my $self = shift;
    if ( $self->{related_numbers} ) {
        return $self->{related_numbers};
    }
    else {
        return;
    }
}

=head1 AUTHOR

Colin Campbell, C<< <colinsc@cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-edifact-interchange at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Business-Edifact-Interchange>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Business::Edifact::Message



=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011,2012 Colin Campbell.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;    # End of Business::Edifact::Message::LineItem


lib/Business/Edifact/Message.pm
package Business::Edifact::Message;

use warnings;
use strict;
use 5.010;
use Carp;
use Business::Edifact::Message::LineItem;

=head1 NAME

Business::Edifact::Message - Class that models Edifact Messages

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

Parses an individual Edifact message
Message objects are instantiated by Business::Edifact::Interchange and an array
of them is returned in its messages function
   $interchange->parse($incoming);
   my $m_array = $interchange->messages();
   for my $msg (@{$m_array}) {
      ...retrieve message data
   }

=cut

=head1 SUBROUTINES/METHODS

=head2 new

Called by Business::Edifact::Interchange to instantiate a new Message
object. The caller passes the header fields with the 
reference number identifier and message type

=cut

sub new {
    my $class      = shift;
    my $hdr_fields = shift;

    my $self = {
        ref_num            => $hdr_fields->[0]->[0],
        message_identifier => $hdr_fields->[1],
        reference          => [],
        addresses          => [],
        lines              => [],
        free_text          => [],
        segment_group      => 0,
        type               => $hdr_fields->[1]->[0],
        segment_handler    => _init_sh(),
    };

    bless $self, $class;
    return $self;
}

sub _init_sh {
    return {
        BGM => \&handle_bgm,
        DTM => \&handle_dtm,
        PAT => \&handle_pat,
        RFF => \&handle_rff,
        CUX => \&handle_cux,
        NAD => \&handle_nad,
        LIN => \&handle_lin,
        PIA => \&handle_pia,
        IMD => \&handle_imd,
        QTY => \&handle_qty,
        GIR => \&handle_gir,
        MOA => \&handle_moa,
        TAX => \&handle_tax,
        ALC => \&handle_alc,
        RTE => \&handle_rte,
        LOC => \&handle_loc,
        PRI => \&handle_pri,
        UNS => \&handle_uns,
        CNT => \&handle_cnt,
        FTX => \&handle_ftx,
    };
}

=head2 add_segment

Process the next data segment

=cut

sub add_segment {
    my $self     = shift;
    my $tag      = shift;
    my $data_arr = shift;

    my $handler = $self->{segment_handler}->{$tag};
    if ($handler) {
        $self->$handler($data_arr);
    }
    return;
}

=head2 type

return the message's type
e.g. 'QUOTES' or 'ORDERS'

=cut

sub type {
    my $self = shift;
    return $self->{type};    # e.g. 'QUOTES'
}

=head2 function

Returns the message's function field
May be 'original' or 'retransmission'

=cut

sub function {
    my $self = shift;
    my $f    = $self->{bgm_data}->[2]->[0];
    if ( $f == 9 ) {
        return 'original';
    }
    elsif ( $f == 7 ) {
        return 'retransmission';
    }
    elsif ( $f == 43 ) {
        return 'additional transmission';
    }
    elsif ( $f == 46 ) {
        return 'provisional';
    }
    else {
        return $f;
    }
}

=head2 message_code

Returns 3 character message code from the BGM message

=cut

sub message_code {
    my $self = shift;
    return $self->{bgm_data}->[0]->[0];
}

=head2 currency_code

=cut

sub currency_code {
    my $self = shift;
    if ( exists $self->{currency} ) {
        return $self->{currency}->[1];
    }
    return;
}

=head2 reference_number

=cut

sub reference_number {
    my $self = shift;
    return $self->{ref_num};
}

=head2 date_of_message

=cut

sub date_of_message {
    my $self = shift;
    return $self->{message_date};
}

=head2 items

return the list of lineitems

=cut

sub items {
    my $self = shift;
    return $self->{lines};
}

=head2 handle_bgm

=cut

sub handle_bgm {
    my ( $self, $data_arr ) = @_;
    $self->{bgm_data} = $data_arr;
    return;
}

=head2 handle_dtm

NB DTM can occur in different segment groups

=cut

sub handle_dtm {
    my ( $self, $data_arr ) = @_;
    my ( $qualifier, $date, $format ) = @{ $data_arr->[0] };
    if ( $self->{segment_group} == 0 ) {    # message header
                                            #TBD standard allows 35 repeats
        if ( $qualifier == 137 ) {
            $self->{message_date} = $date;
        }
        elsif ( $qualifier == 36 ) {
            $self->{expirty_date} = $date;
        }
        elsif ( $qualifier == 131 ) {
            $self->{tax_point_date} = $date;
        }
    }
    elsif ( $self->{segment_group} == 27 ) {
        $self->{lines}->[-1]->addsegment( 'datetimeperiod', $data_arr );
    }
    elsif ( $self->{segment_group} == 11 ) {
        $self->{payment_terms}->{payment_date} = $date;
    }
    return;
}

=head2 handle_pat

=cut

sub handle_pat {    # invoices only
    my ( $self, $data_arr ) = @_;
    if ( $data_arr->[0]->[0] == 1 ) {
        $self->{payment_terms} = {
            type  => 'basic',
            terms => $data_arr->[2],
        };
    }
    elsif ( $data_arr->[0]->[0] == 3 ) {
        $self->{payment_terms} = { type => 'fixed_date', };
    }
    return;
}

=head2 handle_rff

=cut

sub handle_rff {
    my ( $self, $data_arr ) = @_;
    if ( $data_arr->[0]->[0] eq 'VA' ) {
        $self->{supplier_vat_number} = $data_arr->[0]->[1];
    }
    if ( $self->{segment_group} == 0 ) {
        $self->{segment_group}     = 1;    # 1 mandatory occurence
        $self->{message_reference} = {
            qualifier => $data_arr->[0]->[0],
            number    => $data_arr->[0]->[1],
        };
    }
    elsif ( $self->{segment_group} == 11 ) {    # ref to an address (SG12)
        $self->{addresses}->[-1]->{RFF} = {
            qualifier => $data_arr->[0]->[0],
            number    => $data_arr->[0]->[1],
        };

    }
    elsif ( $self->{segment_group} == 27 ) {    # ref to an address (SG12)
        $self->{lines}->[-1]->addsegment( 'item_reference', $data_arr );
    }
    else {
        push @{ $self->{reference} },
          {
            qualifier => $data_arr->[0]->[0],
            number    => $data_arr->[0]->[1],
          };
    }
    return;
}

=head2 handle_cux

=cut

sub handle_cux {
    my ( $self, $data_arr ) = @_;
    if ( $self->{segment_group} == 1 ) {

        $self->{currency}      = $data_arr->[0];
        $self->{segment_group} = 4;
    }
    elsif ( $self->{segment_group} == 11 ) {
        $self->{currency} = $data_arr->[0];
    }
    return;
}

=head2 handle_nad

=cut

sub handle_nad {
    my ( $self, $data_arr ) = @_;
    push @{ $self->{addresses} }, { NAD => $data_arr, };
    $self->{segment_group} = 11;
    return;
}

=head2 handle_lin

=cut

sub handle_lin {
    my ( $self, $data_arr ) = @_;
    $self->{segment_group} = 27;
    my $line = {
        line_number            => $data_arr->[0]->[0],
        action_req             => $data_arr->[1]->[0],
        item_number            => $data_arr->[2]->[0],
        item_number_type       => $data_arr->[2]->[1],
        additional_product_ids => [],
        item_description       => [],
        monetary_amount        => [],
    };
    if ( $data_arr->[3]->[0] ) {
        $line->{sub_line_info} = $data_arr->[3];
    }
    my $lineitem = Business::Edifact::Message::LineItem->new($line);

    push @{ $self->{lines} }, $lineitem;
    return;
}

=head2 handle_pia

=cut

sub handle_pia {
    my ( $self, $data_arr ) = @_;
    $self->{lines}->[-1]->addsegment( 'additional_product_ids', $data_arr );
    return;
}

=head2 handle_imd

=cut

sub handle_imd {
    my ( $self, $data_arr ) = @_;
    if ( $data_arr->[0]->[0] eq 'L' ) {    # only handle text at the moment
        if ( $data_arr->[2]->[4] ) {
            $data_arr->[2]->[3] .= $data_arr->[2]->[4];
        }
        $self->{lines}->[-1]->addsegment(
            'item_description',
            {
                code => $data_arr->[1]->[0],
                text => $data_arr->[2]->[3],
            }
        );
    }
    return;
}

=head2 handle_qty

=cut

sub handle_qty {
    my ( $self, $data_arr ) = @_;
    if ( $self->type eq 'INVOIC' ) {
        my $code = $data_arr->[0]->[0];
        if ( $code == 47 ) {
            $self->{lines}->[-1]->{quantity_invoiced} = $data_arr->[0]->[1];
        }
        elsif ( $code == 11 ) {
            $self->{lines}->[-1]->{place_of_delivery}->[-1]
              ->{split_delivery_quantity} = $data_arr->[0]->[1];
        }
    }
    else {
        if ( !$self->{item_locqty_flag} ) {
            $self->{lines}->[-1]->{quantity} = $data_arr->[0]->[1];
        }
        else {
            $self->{lines}->[-1]->{place_of_delivery}->[-1]->{quantity} =
              $data_arr->[0]->[1];
            delete $self->{item_locqty_flag};
        }
    }
    return;
}

=head2 handle_gir

=cut

sub handle_gir {
    my ( $self, $data_arr ) = @_;
    my $id = shift @{$data_arr};
    my $relnum = { id => $id->[0], };
    for my $d ( @{$data_arr} ) {
        push @{ $relnum->{ $d->[1] } }, $d->[0];
    }

    push @{ $self->{lines}->[-1]->{related_numbers} }, $relnum;
    return;
}

=head2 handle_moa

=cut

sub handle_moa {
    my ( $self, $data_arr ) = @_;
    if ( $self->{segment_group} == 27 ) {
        if ( !$self->{item_alc_flag} ) {
            my $data = shift @{$data_arr};
            my $ma   = {
                qualifier => $data->[0],
                value     => $data->[1],
            };

            push @{ $self->{lines}->[-1]->{monetary_amount} }, $ma;
        }
        else {
            $self->{lines}->[-1]->{item_allowance_or_charge}->[-1]->{amount} =
              $data_arr->[0]->[1];
        }
    }
    else {
        if ( !$self->{item_alc_flag} ) {
            my $data = shift @{$data_arr};
            my $ma   = {
                qualifier => $data->[0],
                value     => $data->[1],
            };
            push @{ $self->{monetary_amount} }, $ma;
        }
        else {
            my $data = shift @{$data_arr};
            my $ma   = {
                qualifier => $data->[0],
                value     => $data->[1],
            };
            push @{ $self->{allowance_or_charge}->[-1]->{amount} }, $ma;
            delete $self->{item_alc_flag};
        }
    }
    return;
}

=head2 handle_tax

=cut

sub handle_tax {
    my ( $self, $data_arr ) = @_;
    if ( $self->{segment_group} == 27 ) {
        if ( !$self->{item_alc_flag} ) {
            my $tax = {
                type_code     => $data_arr->[1]->[0],
                rate          => $data_arr->[4]->[3],
                category_code => $data_arr->[5]->[0],
            };
            push @{ $self->{lines}->[-1]->{item_tax} }, $tax;
        }
        else {
            my $tax = {
                type_code     => $data_arr->[1]->[0],
                rate          => $data_arr->[4]->[3],
                category_code => $data_arr->[5]->[0],
            };
            push
              @{ $self->{lines}->[-1]->{item_allowance_or_charge}->[-1]->{tax}
              }, $tax;
            delete $self->{item_alc_flag};
        }
    }
    else {
        my $tax = {
            type_code     => $data_arr->[1]->[0],
            rate          => $data_arr->[4]->[3],
            category_code => $data_arr->[5]->[0],
        };
        push @{ $self->{tax} }, $tax;
    }
    return;
}

=head2 handle_alc

=cut

sub handle_alc {
    my ( $self, $data_arr ) = @_;
    if ( $self->{segment_group} == 27 ) {
        my $alc = {
            type         => $data_arr->[0]->[0],
            sequence     => $data_arr->[3]->[0],
            service_code => $data_arr->[4]->[0],
        };
        push @{ $self->{lines}->[-1]->{item_allowance_or_charge} }, $alc;
        $self->{item_alc_flag} = 1;
    }
    else {
        my $alc = {
            type         => $data_arr->[0]->[0],
            sequence     => $data_arr->[3]->[0],
            service_code => $data_arr->[4]->[0],
        };
        push @{ $self->{allowance_or_charge} }, $alc;
        $self->{item_alc_flag} = 1;
    }
    return;
}

=head2 handle_rte

=cut

sub handle_rte {
    my ( $self, $data_arr ) = @_;
    if ( $self->{item_alc_flag} == 1 ) {
        $self->{lines}->[-1]->{item_allowance_or_charge}->[-1]->{rate} =
          $data_arr->[0]->[1];
        delete $self->{item_alc_flag};
    }
    return;
}

=head2 handle_loc

=cut

sub handle_loc {
    my ( $self, $data_arr ) = @_;
    if ( $self->{segment_group} == 27 && $data_arr->[0]->[0] == 7 ) {
        my $loc = {
            place     => $data_arr->[1]->[0],
            type_code => $data_arr->[1]->[2],
        };
        push @{ $self->{lines}->[-1]->{place_of_delivery} }, $loc;
        $self->{item_locqty_flag} = 1;
    }
    return;
}

=head2 handle_pri

=cut

sub handle_pri {
    my ( $self, $data_arr ) = @_;
    push @{ $self->{lines}->[-1]->{price} }, {

        # qualifier: AAA = net AAB = gross AAE/F = info
        qualifier            => $data_arr->[0]->[0],
        price                => $data_arr->[0]->[1],
        price_type           => $data_arr->[0]->[2],
        price_type_qualifier => $data_arr->[0]->[3],
    };
    return;
}

=head2 handle_uns

=cut

sub handle_uns {
    my ( $self, $data_arr ) = @_;
    $self->{segment_group} = -1;    # summary does not have a seg group
    return;
}

=head2 handle_cnt

=cut

sub handle_cnt {
    my ( $self, $data_arr ) = @_;
    if ( $data_arr->[0]->[0] == 2 ) {
        $self->{summary_count} = $data_arr->[0]->[1];
    }
    return;
}

=head2 handle_ftx

=cut

sub handle_ftx {
    my ( $self, $data_arr ) = @_;
    my $text_field = {
        qualifier => $data_arr->[0]->[0],
        function  => $data_arr->[1]->[0],
        reference => $data_arr->[2],
    };
    if ( $data_arr->[3] ) {
        $text_field->{text} = join q{ }, @{ $data_arr->[3] };
    }
    if ( @{ $self->{lines} } ) {    # at the lineitem level
        $self->{lines}->[-1]->{free_text} = $text_field;
    }
    else {
        push $self->{free_text}, $text_field;
    }
    return;
}

=head1 AUTHOR

Colin Campbell, C<< <colinsc@cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-edifact-interchange at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Business-Edifact-Interchange>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Business::Edifact::Message


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Colin Campbell.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;
